src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+  def getLearningRate: Double = $(learningRate)
+  def setLearningRate(value: Double): this.type = set(learningRate, value)
+
+  val powerT = new DoubleParam(this, "powerT", "t power value")
+  def getPowerT: Double = $(powerT)
+  def setPowerT(value: Double): this.type = set(powerT, value)
+
+  val l1 = new DoubleParam(this, "l1", "l_1 lambda")
+  def getL1: Double = $(l1)
+  def setL1(value: Double): this.type = set(l1, value)
+
+  val l2 = new DoubleParam(this, "l2", "l_2 lambda")
+  def getL2: Double = $(l2)
+  def setL2(value: Double): this.type = set(l2, value)
+
+  val interactions = new StringArrayParam(this, "interactions", "Interaction terms as specified by -q")
+  def getInteractions: Array[String] = $(interactions)
+  def setInteractions(value: Array[String]): this.type = set(interactions, value)
+
+  val ignoreNamespaces = new Param[String](this, "ignoreNamespaces", "Namespaces to be ignored (first letter only)")
+  def getIgnoreNamespaces: String = $(ignoreNamespaces)
+  def setIgnoreNamespaces(value: String): this.type = set(ignoreNamespaces, value)
+
+  val initialModel = new ByteArrayParam(this, "initialModel", "Initial model to start from")
+  def getInitialModel: Array[Byte] = $(initialModel)
+  def setInitialModel(value: Array[Byte]): this.type = set(initialModel, value)
+
+  val cacheRows = new BooleanParam(this, "cacheRows", "Cache rows in multi-pass mode")
+  setDefault(cacheRows -> false)
+  def getCacheRows: Boolean = $(cacheRows)
+  def setCacheRows(value: Boolean): this.type = set(cacheRows, value)
+
+  val enableCacheFile = new BooleanParam(this, "enableCacheFile", "Enable local cache file")
+  setDefault(enableCacheFile -> false)
+  def getEnableCacheFile: Boolean = $(enableCacheFile)
+  def setEnableCacheFile(value: Boolean): this.type = set(enableCacheFile, value)
+
+  // abstract methods that implementors need to provide (mixed in through Classifier,...)
+  def getLabelCol: String
+  def getFeaturesCol: String
+
+  implicit class ParamStringBuilder(sb: StringBuilder) {
+    def appendParam[T](param: Param[T], option: String): StringBuilder = {
+      if (!get(param).isEmpty) {
+
+        param match {
+          case _: StringArrayParam => {
+            for (q <- get(param).get)
+              sb.append(' ').append(option).append(' ').append(q)
+          }
+          case _ => sb.append(' ').append(option).append(' ').append(get(param).get)
+        }
+      }
+
+      sb
+    }
+  }
+
+  val hashSeed = new IntParam(this, "hashSeed", "Seed used for hashing")
+  setDefault(hashSeed -> 0)
+
+  def getHashSeed: Int = $(hashSeed)
+  def setHashSeed(value: Int): this.type = set(hashSeed, value)
+
+  val numBits = new IntParam(this, "numBits", "Number of bits used")
+  setDefault(numBits -> 18)
+
+  def getNumBits: Int = $(numBits)
+  def setNumBits(value: Int): this.type = set(numBits, value)
+
+  private def createLabelSetter(df: DataFrame) = {
```

Comment:
consider putting all params in a separate params class and putting the base methods in a separate class, I think that would limit a lot of confusion.  Also, a lot of our style checks have limits on number of functions/params per class, so I've had to break even the lightgbm params into separate classes - not necessary here but just another reason to keep the params separate from all of the base methods.

Reply:
since all the params are in the base class we decided to keep it there until the style checks complain.